File: Lectures/10-Pointers-References/cat.cc
#include <string>
#include <cstring>
#include <iostream>
#include "cat.hh"
using namespace std;

Cat::Cat() { 
  string colours[] = {"grey", ""};
  Cat("youthere", colours);
}

Cat::Cat(const char* theName, string* theColours) {
  setName(theName);
  copyColours(theColours);
  cout << "Created " << toString() << endl;
}

Cat::Cat(const Cat& original) { // Could break up original and pass on to the other constructor  : Cat(original.myName, original.theColours)
  setName(original.myName);
  copyColours(original.myColours);
  cout << "Copy of " << original.toString() << endl;
}

void Cat::setName(const char* theName) {
  char* oldName = myName;
  myName = new char[strlen(theName+1)];
  strcpy(myName, theName);

  if (0 != oldName) {
    cout << oldName << " is now known as " << myName << endl;
  }
}

string Cat::toString(void) const {
  return string("a cat by the name ") + myName + " with the colours " + listColours();
}

string Cat::listColours(void) const {
  string out;
  int pos = 0;

  while ("" != myColours[pos]) {
    out += myColours[pos] + ", ";
    pos++;
  }

  return out;
}

void Cat::copyColours(const string* theColours) {
  int length=0;
  while ("" != theColours[length++]) {}
  myColours = new string[length];

  int pos=0;
  do {
    myColours[pos] = theColours[pos];
  } while ("" != theColours[pos++]); // Sometimes we want to do the loop check at the end: do {} while()
}

File: Lectures/10-Pointers-References/Cat.java
import java.util.ArrayList;
public class Cat {
  public String name;
  public ArrayList<String> colours;
  public final int x = 0;

  Cat() { // Default constructor
    colours = new ArrayList<>();
    colours.add("Brown");
    name = "youthere";
    System.out.println("Default Constructor " + name);
  }

  Cat(String theName, ArrayList<String> theColours) {
    name = theName;
    colours = theColours;
    System.out.println("Constructor " + name);
  }

  Cat(Cat original) { // Copy constructor
    name = original.name; // Strings are special, so this one behaves as expected
    colours = original.colours; // But be careful with other object references. See in the main function below...
    System.out.println("Copy Constructor " + name);
  }

  public String toString() {
    return "a cat with the name " + name + " and the colours " +colours.toString();
  }

  public static void doSomething(Cat theCat) {
    System.out.println("Doing something with " + theCat.toString());
  }

  public static void main(String [] args) {
    Cat stray = new Cat();
    System.out.println("Stray: " + stray.toString());

    //Cat mongrel = stray; // Create a new reference 'mongrel', and points it at the same object as 'stray'
    Cat mongrel = new Cat(stray); // Invoke the copy constructor

    mongrel.name="Manny";
    mongrel.colours.add("White"); // This is where the "Just copy the reference" will bite us!
    System.out.println("Mongrel: " + mongrel.toString());
    System.out.println("Stray, as it is now: " + stray.toString());

    doSomething(mongrel); // Will only pass the reference, less risk of accidentally copying the object
  }
}